### PR TITLE
feat(nestjs-query): add support to custom get requests

### PR DESCRIPTION
**Pull request by igordonin**
## PR Checklist

Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://refine.dev/docs/guides-concepts/contributing/#commit-convention

## Bugs / Features

- [x] Related issue(s) linked
- [x] Tests for the changes have been added
- [x] Docs have been added / updated
- [x] Changesets have been added https://refine.dev/docs/guides-concepts/contributing/#creating-a-changeset

## What is the current behavior?

Custom requests do not make use of the params that are passed to it, including the HTTP Request Method, the URL and the headers.

## What is the new behavior?

Now it uses all of the custom params and correctly supports POST and GET Request Methods, throws errors for unsupported methods, and also uses the custom URL and headers passed to the function.

## Notes for reviewers

Not at this time

### CODE REVIEW

File: packages/nestjs-query/src/dataProvider/index.ts [hunk 1]
```diff
@@ -419,30 +419,38 @@ const dataProvider = (client: GraphQLClient): Required<DataProvider> => {
...
 
       if (meta?.rawQuery) {
-        const response = await client.request<BaseRecord>(
-          meta.rawQuery,
-          meta.variables,
-        );
+        const response = await _client.request<BaseRecord>({
+          document: meta.rawQuery,
+          variables: meta.variables,
+          requestHeaders: headers,
```
Comment (igordonin) [inline packages/nestjs-query/src/dataProvider/index.ts:452]: Oops, those aren't needed
```suggestion
          requestHeaders: headers,
```

File: packages/nestjs-query/src/dataProvider/index.ts [hunk 2]
```diff
@@ -472,7 +480,11 @@ const dataProvider = (client: GraphQLClient): Required<DataProvider> => {
             variables = gqlMutation.variables;
           }
 
-          const response = await client.request<BaseRecord>(query, variables);
+          const response = await _client.request<BaseRecord>({
+            document: query,
+            variables,
+            requestHeaders: headers,
```
Comment (igordonin) [inline packages/nestjs-query/src/dataProvider/index.ts:486]: Oops, those aren't needed
```suggestion
          requestHeaders: headers,
```

File: packages/nestjs-query/src/dataProvider/index.ts [hunk 3]
```diff
@@ -419,30 +419,38 @@ const dataProvider = (client: GraphQLClient): Required<DataProvider> => {
...
 
       if (gqlOperation) {
-        const response: any = await client.request(
-          gqlOperation,
-          meta?.variables ?? {},
-        );
+        const response: any = await _client.request<BaseRecord>({
+          document: gqlOperation,
+          variables: meta?.variables,
+          requestHeaders: headers,
```
Comment (igordonin) [inline packages/nestjs-query/src/dataProvider/index.ts:442]: Oops, those aren't needed
```suggestion
          requestHeaders: headers,
```

**Review by BatuhanW — COMMENTED** [review]
Hey @igordonin thanks for the PR! I'll do a more in-depth review later but, wanted to mention that we should keep existing POST request mocks and tests instead of replacing them with GET. Better to have both of them.

**Comment by igordonin** [comment]
> Hey @igordonin thanks for the PR! I'll do a more in-depth review later but, wanted to mention that we should keep existing POST request mocks and tests instead of replacing them with GET. Better to have both of them.

Np, Glad to help.

Agreed. We still have both though. The ones I replaced were being used to test GET requests. I have not removed any of the previous tests doing POST requests.

**Comment by BatuhanW** [comment]
> > Hey @igordonin thanks for the PR! I'll do a more in-depth review later but, wanted to mention that we should keep existing POST request mocks and tests instead of replacing them with GET. Better to have both of them.
> 
> Np, Glad to help.
> 
> Agreed. We still have both though. The ones I replaced were being used to test GET requests. I have not removed any of the previous tests doing POST requests.

I see, I was confused by the diff and poor test case naming!

We also wanted to allow users to make REST requests if they need to, for cases like custom authentication or one off requests like PATCH, PUT into endpoints. Do you think we could do that in this custom method as well?

**Comment by igordonin** [comment]
> We also wanted to allow users to make REST requests if they need to, for cases like custom authentication or one off requests like PATCH, PUT into endpoints. Do you think we could do that in this custom method as well?

Not as is, no. It's currently using `gql-request` to make the requests, so it only supports `GET` and `POST`. So this whole DataProvider is currently coupled to GraphQL not only because of the dependencies but also because of the implementations.

I don't think you should try to provide a single generic DataProvider though...

**Comment by BatuhanW** [comment]
Hey @igordonin looks great! I've discussed with core team and we've realized an edge-case that can be handled here.

https://github.com/refinedev/refine/blob/71ee737c2bfd89f928546d66d2f78b1b4cc1c544/examples/app-crm/src/providers/data/index.ts#L15

In `app-crm` example, we are overriding `fetch` method of gql client to add refresh token mechanism.

I think we can do something like this in custom method:

```
const _client = new GraphQLClient(url, {...client.requestConfig, headers })
```

This way, (hopefully) we can preserve user's client configuration.

**Comment by igordonin** [comment]
Ok, done that. Now we keep the middlewares and override only headers and http method.

File: packages/nestjs-query/src/dataProvider/index.ts [hunk 4]
```diff
@@ -419,30 +419,37 @@ const dataProvider = (client: GraphQLClient): Required<DataProvider> => {
...
+      if (!SUPPORTED_METHODS.some((it) => it === method)) {
+        throw Error(`GraphQL does not support ${method} method.`);
       }
 
+      const validMethod = method as "get" | "post";
+
+      const _client = new GraphQLClient(requestUrl, {
+        ...client.requestConfig,
+        method: validMethod,
+        headers,
```
Comment (BatuhanW) [inline packages/nestjs-query/src/dataProvider/index.ts:434]:
```suggestion
        headers: { ..client.requestConfig.headers, ...headers },
```

I think would be nice to not lose headers, if there are any.

**Review by BatuhanW — COMMENTED** [review]
Great PR! Thanks for the updates, added a minor comment. We'll release it with August release.

File: packages/nestjs-query/src/dataProvider/index.ts (hunk 4 above)
Comment (igordonin) [reply packages/nestjs-query/src/dataProvider/index.ts:434]: nice catch

**Review by BatuhanW — APPROVED** [review]
Thanks for the great PR! Great stuff. 🚀

**Pull request merged**